### DISCUSSION TITLE
pickle with prepend/keep_alive extras

### DISCUSSION THREAD

**Discussion by adedner** [pybind/pybind11#4360] [Ideas]
Hi.
First of all thanks for the great project, has simplified getting my students to use our C++ code a lot!

I recently started on adding pickling support and ran into two problems (I opened two issues):

__Issue__: 
I have a default pickling method that is automatically registered and would like to override that in certain cases.

__Description__:
Adding a second `def(pybind11::pickling...)` does not work since the previous version is sued. Adding `pybind11::prepend()`  leads to unexpected results - the default getstate` is used together with the new `setstate`. The reason is that in the `pickle_factory` the `extra` argument is only added to the `setter`. I can't find a workaround but was wondering if in this case having multiple overloads for `setstate/getstate` makes sense? Possibly a new `def(pickle(...))` should remove the first one or always prepend (that was easy enough to do)?

__Solution__:
At least failing if `prepend` is used in the extra argument would probably avoid some unexpected behavior.
What worked for me was testing for `pybind11:prepend` in the `Extra` variadic template when calling `pickle_factory::execute` and if present adding that to the `__getstate__` registration.
I can provide a MR and possibly a test with that approach if it makes sense.
But as mentioned above perhaps simply removing a previous registration would be better since I believe that overloads would not be called anyway - I'm not that far down the rabbit hole yet to implement this approach.

**Comment by gruenich** [comment]
I think your request is reasonable. I prefer the way to remove a pickling instead of having multiple overloads.
